userspace/libscap/engine/savefile/scap_reader_buffered.c

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright (C) 2022 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#include "scap_reader.h"
+#include <string.h>
+
+typedef struct reader_handle
+{
+    bool m_free_reader; ///< Whether the reader should be free-d on close
+    bool m_has_err; ///< True if the most recent m_reader operation had an error
+    uint8_t* m_buffer; ///< The buffer used to read data from m_reader
+    uint32_t m_buffer_cap; ///< The physical size of the buffer
+    uint32_t m_buffer_len; ///< The number of bytes used in the buffer
+    uint32_t m_buffer_off; ///< The cursor position in the buffer
+    int64_t m_offset; ///< The cursor position in the underlying reader
+    scap_reader_t* m_reader; ///< The reader to read from in buffered mode
+} reader_handle_t;
+
+static int buffered_read(scap_reader_t *r, void* buf, uint32_t len)
+{
+    ASSERT(r != NULL);
+    reader_handle_t* h = (reader_handle_t*) r->handle;
+    uint8_t* buf_bytes = (uint8_t*) buf;
+    uint32_t size = 0;
+    while (len > 0 && !h->m_has_err)
+    {
+        if (h->m_buffer_off >= h->m_buffer_len)
+        {
+            int nread = h->m_reader->read(h->m_reader, h->m_buffer, h->m_buffer_cap);
+            if (nread <= 0)
+            {
+                // invalidate next read
+                h->m_has_err = true;
+                return buf_bytes - (uint8_t*) buf;
+            }
+            h->m_offset += nread;
+            h->m_buffer_off = 0;
+            h->m_buffer_len = (uint32_t) nread;
+        }
+        size = len < (h->m_buffer_len - h->m_buffer_off) ? len : (h->m_buffer_len - h->m_buffer_off);
```

Comment:
Can we split `h->m_buffer_len - h->m_buffer_off` into a separate local variable and use that in this line? I think it would improve readability, something like:
```c
uint32_t buffer_len = h->m_buffer_len - h->m_buffer_off;
size = len < buffer_len ? len : buffer_len;
```
WDYT?